src/pages/login/index.tsx

```diff
@@ -1,67 +1,134 @@
+  confirmPasswordError: '',
+  showPasswordModal: false,
+  toastMessage: '',
+  showingToast: false,
+  errorMessage: '',
+  toastType: 'success' as 'create' | 'delete' | 'success' | 'info',
+}
 
-// const GOOGLE_CLIENT_ID = process.env.NEXT_PUBLIC_GOOGLE_CLIENT_ID
-// const KAKAO_CLIENT_ID = process.env.NEXT_PUBLIC_KAKAO_CLIENT_ID
-// const BASE_URL = process.env.NEXT_PUBLIC_BASE_URL
+const reducer = (state: typeof initialState, action: SignupAction) => {
+  switch (action.type) {
+    case 'SET_EMAIL':
+      return { ...state, email: action.payload }
+    case 'SET_PASSWORD':
+      return { ...state, password: action.payload }
+    case 'SET_EMAIL_ERROR':
+      return { ...state, emailError: action.payload }
+    case 'SET_PASSWORD_ERROR':
+      return { ...state, passwordError: action.payload }
+    case 'SET_MODAL':
+      return {
+        ...state,
+        showPasswordModal: action.payload.show,
+        errorMessage: action.payload.message,
+      }
+    case 'SET_TOAST':
+      return {
+        ...state,
+        toastMessage: action.payload.message,
+        toastType: action.payload.type,
+        showingToast: true,
+      }
+    case 'HIDE_TOAST':
+      return {
+        ...state,
+        showingToast: false,
+        toastMessage: '',
+        toastType: 'success',
+      }
+    case 'TOGGLE_PASSWORD_VISIBILITY':
+      return { ...state, showPassword: !state.showPassword }
+    default:
+      return state
+  }
+}
 
 export default function Login() {
-  const [email, setEmail] = useState('')
-  const [password, setPassword] = useState('')
-  const [showPassword, setShowPassword] = useState(false)
-  const [emailError, setEmailError] = useState('')
-  const [passwordError, setPasswordError] = useState('')
-  //
-
-  // const handleGoogleLogin = () => {
-  //   const redirectUri = `${process.env.NEXT_PUBLIC_BASE_URL}/auth/google`
+  const [state, dispatch] = useReducer(reducer, initialState)
+  const router = useRouter()
+  const { setAccessToken } = useAuthStore()
 
-  //   window.location.href = `https://accounts.google.com/o/oauth2/v2/auth?client_id=${process.env.NEXT_PUBLIC_GOOGLE_CLIENT_ID}&redirect_uri=${redirectUri}&response_type=code&scope=profile email`
-  // }
+  // useFormSignup 훅 호출
+  useFormSignup(state, dispatch) // 이메일, 비밀번호 유효성 검사
 
-  // const handleKakaoLogin = () => {
-  //   window.location.href = `https://kauth.kakao.com/oauth/authorize?client_id=${KAKAO_CLIENT_ID}&redirect_uri=${BASE_URL}/auth/kakao&response_type=code`
-  // }
-  //
   const isFormValid = useMemo(() => {
-    return /^[^\s@]+@[^\s@]+\.[^\s@]+$/.test(email) && password.length >= 8
-  }, [email, password])
-
-  const { setAccessToken } = useAuthStore() // 토큰 관리
-  const router = useRouter()
+    return (
+      /^[^\s@]+@[^\s@]+\.[^\s@]+$/.test(state.email) &&
+      state.password.length >= 8 &&
+      !state.emailError &&
+      !state.passwordError
+    )
+  }, [state.email, state.password, state.emailError, state.passwordError])
 
```

Comment:
form vaildation을 하나의 값으로 분리하기보다 각 필드별로 분리하고, 모든 필드의 validation결과값을 바탕으로 만들면 코드를 이해하기가 더 좋아보여요.
그리고 useMemo는 사용하지 않아도 될 것 같습니다! 참고로 useMemo의 사용케이스는 연산이 복잡한 값의 경우 메모이제이션을(캐싱) 하기 위함입니다.
사용하신 케이스는 연산이 복잡하지 않기 때문에 불필요한 메모이제이션 비용이 더 비효율적인것 같아요.


```suggestion
  const isEmailValid = /^[^\s@]+@[^\s@]+\.[^\s@]+$/.test(state.email);
  const isPasswordValid = state.password.length >= 8;
  const isFormValid = isEmailValid && isPasswordValid && !state.emailError && !state.passwordError; // << emailError, passwordError 두개 state도 is**Valid 에 녹일 수 있을 것 같네요.

```